src/io/import_off.cc

```diff
@@ -2,24 +2,234 @@
+    return std::make_unique<PolySet>(3);
+  }
+
+  PRINTDB("%d vertices, %d faces, %d edges.", vertices_count % faces_count % edges_count);
+
+  auto ps = std::make_unique<PolySet>(3);
+  ps->vertices.reserve(vertices_count);
+  ps->indices.reserve(faces_count);
+
+  while ((!f.eof()) && (vertex++ < vertices_count)) {
+    if (!getline_clean("reading vertices: end of file"))
+      return std::make_unique<PolySet>(3);
+    if (line.length() == 0)
+      continue;
+
+    boost::trim(line);
+    boost::split(words, line, boost::is_any_of(" \t"), boost::token_compress_on);
+    if (words.size() < 3) {
+      AsciiError("can't parse vertex: not enough data");
+      return std::make_unique<PolySet>(3);
+    }
+
+    try {
+      Vector3d v;
+      for (int i = 0; i < 3; i++) {
+        v[i]= boost::lexical_cast<double>(words[i]);
+      }
+      //PRINTDB("Vertex[%ld] = { %f, %f, %f }", vertex % v[0] % v[1] % v[2]);
+      int o = dimension;
+      if (has_normals) {
+        ; // TODO words[o++]
+        o += 0;
+      }
+      if (has_color) {
+        ; // TODO: Meshlab appends color there, probably to allow gradients
+        o += 3; // 4?
+      }
+      if (has_textures) {
+        ; // TODO
+      }
+      ps->vertices.push_back(v);
+    } catch (const boost::bad_lexical_cast& blc) {
+      AsciiError("can't parse vertex: bad data");
+      return std::make_unique<PolySet>(3);
+    }
+  }
+
+  while (!f.eof() && (face++ < faces_count)) {
+    if (!getline_clean("reading faces: end of file"))
+      return std::make_unique<PolySet>(3);
+    if (line.length() == 0)
+      continue;
+
+    boost::trim(line);
+    boost::split(words, line, boost::is_any_of(" \t"), boost::token_compress_on);
+    if (words.size() < 1) {
+      AsciiError("can't parse face: not enough data");
+      return std::make_unique<PolySet>(3);
+    }
+
+    try {
+      unsigned long n=boost::lexical_cast<unsigned long>(words[0]);
+      if (words.size() - 1 < n) {
+        AsciiError("can't parse face: missing indices");
+        return std::make_unique<PolySet>(3);
+      }
+      ps->indices.emplace_back().reserve(n);
+      //PRINTDB("Index[%d] [%d] = { ", face % n);
+      for (int i = 0; i < n; i++) {
+        int ind=boost::lexical_cast<int>(words[i+1]);
+        //PRINTDB("%d, ", ind);
+        if(ind >= 0 && ind < vertices_count)
+          ps->indices.back().push_back(ind);
+      }
+      //PRINTD("}");
+      if (words.size() >= n + 4) {
+        // TODO: handle optional color info
+        /*
+        int r=boost::lexical_cast<int>(words[n+1]);
+        int g=boost::lexical_cast<int>(words[n+2]);
+        int b=boost::lexical_cast<int>(words[n+3]);
+        */
+      }
+    } catch (const boost::bad_lexical_cast& blc) {
+      AsciiError("can't parse face: bad data");
+      return std::make_unique<PolySet>(3);
+    }
+  }
+
+  //PRINTDB("PS: %ld vertices, %ld indices", ps->vertices.size() % ps->indices.size());
+  f.close();
```

Comment:
close is unnecessary; the destructor will close.

Reply:
OK